subprojects/logging/src/main/java/org/gradle/internal/logging/console/LogGroupingOutputEventListener.java

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.logging.console;
+
+import com.google.common.collect.Lists;
+import org.gradle.api.Nullable;
+import org.gradle.api.logging.LogLevel;
+import org.gradle.internal.SystemProperties;
+import org.gradle.internal.logging.events.BatchOutputEventListener;
+import org.gradle.internal.logging.events.CategorisedOutputEvent;
+import org.gradle.internal.logging.events.EndOutputEvent;
+import org.gradle.internal.logging.events.LogEvent;
+import org.gradle.internal.logging.events.OperationIdentifier;
+import org.gradle.internal.logging.events.OutputEvent;
+import org.gradle.internal.logging.events.ProgressCompleteEvent;
+import org.gradle.internal.logging.events.ProgressEvent;
+import org.gradle.internal.logging.events.ProgressStartEvent;
+import org.gradle.internal.logging.events.RenderableOutputEvent;
+import org.gradle.internal.logging.events.StyledTextOutputEvent;
+import org.gradle.internal.logging.text.StyledTextOutput;
+import org.gradle.internal.progress.BuildOperationType;
+import org.gradle.internal.time.TimeProvider;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+
+public class LogGroupingOutputEventListener extends BatchOutputEventListener {
+    static final int SCHEDULER_CHECK_PERIOD_MS = 5000;
+    public static final String EOL = SystemProperties.getInstance().getLineSeparator();
+    private final BatchOutputEventListener listener;
+    private final ScheduledExecutorService executor;
+    private final TimeProvider timeProvider;
+    private final Object lock = new Object();
+
+    // Allows us to map progress and complete events to start events
+    private final Map<OperationIdentifier, Object> progressIdToBuildOperationIdMap = new LinkedHashMap<OperationIdentifier, Object>();
+
+    // Maintain a hierarchy of all build operation ids, so that child operations are grouped under their ancestor
+    private final Map<Object, Object> buildOperationIdForest = new LinkedHashMap<Object, Object>();
```

Comment:
Could we potentially break out the management of this Map into a separate class to make the logic more readable.